src/finalizer/utils/helios.ts

```diff
@@ -757,3 +770,31 @@ function addUpdateOnlyTxn(
     destinationChainId: l2ChainId,
   });
 }
+
+/**
+ *
+ * @notice This function ensures that there's a match between `ZK API.vkey` _and_ `SP1Helios.vkey`. If these two don't
+ * match, the generated proof cannot be used to confirm messages on the destination chain.
+ *
+ * When upgrading the ZK Helios setup, ZK API will always have to be upgraded _after_ we send an upgrade message to all
+ * the V4 chains, because ZK API has to _first_ generate all the proofs for "upgrade message" to get confirmed on the
+ * destination chain. This means that ELF vkey might become stale in the ZK API (say some contract already upgraded but
+ * we didn't yet re-deploy the ZK API with the new ELF). This check eases the operational overhead to the upgrade by not
+ * allowing to request a proof with an incorrect(stale) ELF vkey.
+ */
+async function ensureVkeysMatch(apiBaseUrl: string, sp1Helios: ethers.Contract): Promise<void> {
+  const [apiResp, contractVkeyRaw] = await Promise.all([
+    axios.get<VkeyResponse>(`${apiBaseUrl}/v1/api/vkey`),
+    sp1Helios.heliosProgramVkey(),
+  ]);
+
+  const apiVkeyRaw = apiResp?.data?.vkey;
+  const apiVkey = apiVkeyRaw !== undefined ? apiVkeyRaw.toLowerCase() : undefined;
+  const contractVkey = contractVkeyRaw !== undefined ? contractVkeyRaw.toLowerCase() : undefined;
```

Comment:
Minor driveby comment - you could consider this to avoid having to repeat the variable names, which can be error-prone:

```suggestion
  const apiVkey = apiVkeyRaw?.toLowerCase();
  const contractVkey = contractVkeyRaw?.toLowerCase();
```